diff --git a/include/FreeRTOS.h b/include/FreeRTOS.h
index f31e72abb..6a552f567 100644
--- a/include/FreeRTOS.h
+++ b/include/FreeRTOS.h
@@ -2234,6 +2234,14 @@
     #define traceRETURN_xTaskGetSchedulerState( xReturn )
 #endif

+#ifndef traceENTER_xTaskGetSchedulerStateFromISR
+    #define traceENTER_xTaskGetSchedulerStateFromISR()
+#endif
+
+#ifndef traceRETURN_xTaskGetSchedulerStateFromISR
+    #define traceRETURN_xTaskGetSchedulerStateFromISR( xReturn )
+#endif
+
 #ifndef traceENTER_xTaskPriorityInherit
     #define traceENTER_xTaskPriorityInherit( pxMutexHolder )
 #endif
diff --git a/include/task.h b/include/task.h
index 81921320e..7f7ed0a57 100644
--- a/include/task.h
+++ b/include/task.h
@@ -3629,6 +3629,14 @@ void vTaskMissedYield( void ) PRIVILEGED_FUNCTION;
  */
 BaseType_t xTaskGetSchedulerState( void ) PRIVILEGED_FUNCTION;

+/*
+ * Same as xTaskGetSchedulerState but to be used inside ISR
+ * Returns the scheduler state as taskSCHEDULER_RUNNING,
+ * taskSCHEDULER_NOT_STARTED or taskSCHEDULER_SUSPENDED.
+ */
+
+BaseType_t xTaskGetSchedulerStateFromISR( void ) PRIVILEGED_FUNCTION;
+
 /*
  * Raises the priority of the mutex holder to that of the calling task should
  * the mutex holder have a priority less than the calling task.
diff --git a/portable/ThirdParty/Community-Supported-Ports b/portable/ThirdParty/Community-Supported-Ports
index bae4c7aa1..9c5bad7b2 160000
--- a/portable/ThirdParty/Community-Supported-Ports
+++ b/portable/ThirdParty/Community-Supported-Ports
@@ -1 +1 @@
-Subproject commit bae4c7aa19009825ba48071a8fe25dcb8be84880
+Subproject commit 9c5bad7b2f42b4e8420f12290b9bc085d3309e8b
diff --git a/portable/ThirdParty/Partner-Supported-Ports b/portable/ThirdParty/Partner-Supported-Ports
index abc22103e..dc3afc6e8 160000
--- a/portable/ThirdParty/Partner-Supported-Ports
+++ b/portable/ThirdParty/Partner-Supported-Ports
@@ -1 +1 @@
-Subproject commit abc22103e1e6634b33457d4127bff1ab62f27f90
+Subproject commit dc3afc6e837426b4bda81bbb6cf45bfb6f34c7e9
diff --git a/tasks.c b/tasks.c
index 2b7d31198..4613b3921 100644
--- a/tasks.c
+++ b/tasks.c
@@ -6642,6 +6642,42 @@ static void prvResetNextTaskUnblockTime( void )
         return xReturn;
     }

+    BaseType_t xTaskGetSchedulerStateFromISR( void )
+    {
+        BaseType_t xReturn;
+        UBaseType_t uxSavedInterruptStatus;
+
+        traceENTER_xTaskGetSchedulerStateFromISR();
+
+        if( xSchedulerRunning == pdFALSE )
+        {
+            xReturn = taskSCHEDULER_NOT_STARTED;
+        }
+        else
+        {
+            #if ( configNUMBER_OF_CORES > 1 )
+                uxSavedInterruptStatus = taskENTER_CRITICAL_FROM_ISR();
+            #endif
+            {
+                if( uxSchedulerSuspended == ( UBaseType_t ) 0U )
+                {
+                    xReturn = taskSCHEDULER_RUNNING;
+                }
+                else
+                {
+                    xReturn = taskSCHEDULER_SUSPENDED;
+                }
+            }
+            #if ( configNUMBER_OF_CORES > 1 )
+                taskEXIT_CRITICAL_FROM_ISR(uxSavedInterruptStatus);
+            #endif
+        }
+
+        traceRETURN_xTaskGetSchedulerStateFromISR( xReturn );
+
+        return xReturn;
+    }
+
 #endif /* ( ( INCLUDE_xTaskGetSchedulerState == 1 ) || ( configUSE_TIMERS == 1 ) ) */
 /*-----------------------------------------------------------*/

